[Merged by Bors] - Use OS file locks in validator client

michaelsproul opened the pull request:
## Issue Addressed

Closes #1823

## Proposed Changes

* Use OS-level file locking for validator keystores, eliminating problems with lockfiles lingering after ungraceful shutdowns (`SIGKILL`, power outage). I'm using the `fs2` crate because it's cross-platform (unlike `file-lock`), and it seems to have the most downloads on crates.io.
* Deprecate + disable `--delete-lockfiles` CLI param, it's no longer necessary
* Delete the `validator_dir::Manager`, as it was mostly dead code and was only used in the `validator list` command, which has been rewritten to read the validator definitions YAML instead.

## Additional Info

Tested on:

- [x] Linux
- [x] macOS
- [x] Docker Linux
- [x] Docker macOS
- [ ] Windows

michaelsproul commented:
Confirmed working on macOS

blacktemplar reviewed (commented):
Only one small comment/question, otherwise lgtm.

michaelsproul commented on common/lockfile/src/lib.rs:
```diff
@@ -0,0 +1,38 @@
+
+impl Drop for Lockfile {
+    fn drop(&mut self) {
+        let _ = fs::remove_file(&self.path);
```
Ah that's a good point. I know closing the file descriptor does, which happens shortly after this call anyway, but it might be prudent to call `unlock` as well. I'll investigate.

michaelsproul replied:
I've decided not to change anything here for now, in the interests of keeping things simple and not introducing more potential race conditions. I realised if we added an `unlock` before the delete, then the following sequence would be possible:

```
1. Process A unlocks file
2. Process B locks file
3. Process A deletes file
4. ???
```

pawanjay176 commented on common/lockfile/src/lib.rs:
```diff
@@ -0,0 +1,38 @@
+    /// Obtain an exclusive lock on the file at `path`, creating it if it doesn't exist.
+    pub fn new(path: PathBuf) -> Result<Self, LockfileError> {
+        let file = File::create(&path).map_err(LockfileError::UnableToCreateFile)?;
+        file.try_lock_exclusive()
```
Can `try_lock_exclusive` also return an error for any other reason(e.g. no file permission) than the file being locked?

michaelsproul replied:
It can. It's hard to test for because of the concurrency, but I've added a `match` to this `map_err` to make sure we only report the `FileLocked` error when the file is _actually_ locked.

pawanjay176 reviewed (approved):
LGTM! just a minor question

michaelsproul commented:
Ready for merging IMO. Will test Windows support once that's closer to being supported in the rest of Lighthouse.

paulhauner commented on validator_client/src/initialized_validators.rs:
```diff
@@ -226,15 +189,15 @@ impl InitializedValidator {
                     .ok_or_else(|| Error::BadVotingKeystorePath(voting_keystore_path.clone()))?;
 
-                create_lock_file(&voting_keystore_lockfile_path, delete_lockfiles, &log)?;
+                let voting_keystore_lockfile = Lockfile::new(lockfile_path)?;
```
Regarding backwards compat with existing versions of Lighthouse and Teku who also [use our existing strategy](https://github.com/sigp/lighthouse/issues/1823#issuecomment-716232791).

I understand with this PR we'd be unsafe (2 processes 1 file) if old-Lighthouse or Teku have created a lockfile. I guess this isn't a *huge* deal before genesis, but could possibly end up with a slashing.

One potential solution would be to add a `file_existed` field to `Lockfile` and log a `warn!` here if it is ``true`. It might be a handy little reminder to people that their node isn't exiting gracefully, as well.

The above solution doesn't ensure full safety, we'd have to basically revert the whole purpose of this PR achieve that (i.e., stop rebooting after a power failure).

michaelsproul replied:
As discussed out of band we'll let this compatibility with Teku slip for now, as it's quite an obscure benefit (it only applies if the clients are sharing keystore directories, which they usually won't).

And "old-Lighthouse" nodes are still safe from slashing because they will have to use the same slashing protection DB, and there's an exclusive lock there too.

paulhauner commented on account_manager/src/validator/list.rs:
```diff
@@ -1,24 +1,28 @@
 
 pub fn cli_app<'a, 'b>() -> App<'a, 'b> {
-    App::new(CMD).about("Lists the names of all validators.")
+    App::new(CMD).about("Lists the public keys of all validators.")
```
This is not strictly true, since it lists directory names.

michaelsproul replied:
It used to... maybe this is a bit sneaky on my part, but I've changed it so that it reads the `validator_definitions.yaml` and prints the pubkeys themselves. I think that's a bit cleaner than the directory name thing we were doing before, and allows us to delete the `validator_dir::Manager` which was only used here.

michaelsproul commented:
I've added a warning about existing keystores and cleaned up the `ValidatorDir` locks a little more.

Just planning to test under Docker (on macOS and Linux) as per @ajsutton's recommendation (thanks!) then should be good to go.

paulhauner reviewed (approved):
Great! This will make users very happy, I think.

michaelsproul commented:
OK, tested on Docker Linux and Docker macOS. Hopefully CI succeeds with that typo fixed, so...

bors r+

michaelsproul commented:
I'm feeling optimistic about that fix:

bors r+

michaelsproul commented:
Cancelled the regular build to let Bors handle it

paulhauner commented:
https://github.com/sigp/lighthouse/pull/1958/commits/a8b187bb4c899dea0f56559a98e15d21bff8945d looks good to me.

michaelsproul commented:
Damn it, Bors crashed :(

```
{:timeout,
 {GenServer, :call,
  [
    BorsNG.GitHub,
    {:post_commit_status,
     {{:installation, 10233225}, 139952333},
     {"a8b187bb4c899dea0f56559a98e15d21bff8945d", :running,
      "Running",
      "https://app.bors.tech/repositories/26362/log#batch-111416"}},
    10000
  ]}}
```

Time to try again:

bors r-
bors r+
